python/paddle/distributed/auto_parallel/interface.py

```diff
@@ -0,0 +1,352 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+import paddle.fluid.core as core
+import paddle
+from paddle.fluid.framework import Variable
+from paddle.fluid.framework import in_dygraph_mode
+
+__all__ = []
+
+_g_process_mesh_map = dict()
+
+
+def _append_attr_suffix(name):
+    """
+    Append Auto Parallel Suffix for distributed attribute.
+    """
+    return name + core.kAutoParallelSuffix()
+
+
+def _remove_attr_suffix(name):
+    """
+    Remove Auto Parallel Suffix for distributed attribute.
+    """
+    return name.strip(core.kAutoParallelSuffix())
+
+
+def _static_mode_check():
+    if in_dygraph_mode():
+        raise RuntimeError("Now auto-parallel only supports static mode, please"
+                           " use paddle.enable_static().")
+
+
+class ProcessMesh(object):
+    """
+    A mesh is an n-dimensional array of logical processes. The shape of the 
+    n-dimensional array represents the topology of logical processes and each
+    element of the n-dimensional array represent a logical process. For
+    example, the following diagram is represented by the n-dimensional array
+    numpy.array([[2, 4, 5], [0, 1, 3]]), and first logical process is the one
+    with id=2.
+    
+    -------------
+    | 2 | 4 | 5 |
+    -------------
+    | 0 | 1 | 3 |
+    -------------
+
+    Args:
+        mesh (numpy.ndarray): an n-dimensional array of processes. Its data type is int.
+        parent (ProcessMesh): the parent ProcessMesh. None means
+            no parent ProcessMesh.
+    
+    Returns:
+        None
+
+    Raises:
+        ValueError: if ``mesh`` is not an instance of numpy.ndarray.
+
+    Examples:
+        .. code-block:: python
+
+            import numpy as np
+            import paddle
+            import paddle.distributed as dist
+            
+            paddle.enable_static()
+            
+            mesh = dist.ProcessMesh(np.array([[2, 4, 5], [0, 1, 3]]))
+            assert mesh.parent is None
+            assert mesh.topology == [2, 3]
+            assert mesh.process_group == [2, 4, 5, 0, 1, 3]
+    """
+
+    def __init__(self, mesh, parent=None):
```

Comment:
We should make sure the item of mesh is no less than 0 and less than the total used GPUs.

Reply:
done.